README.md
# isoVec

## Description

isoVec is a framework to build complex mixtures of materials and get atomic (mole), weight or volume percentages of each of its substances down to the isotopic composition (the isotope vector, hence the name).
The information on relative atomic weights of isotopes, as well as their abundance in natural occuring elements, is taken from "*Atomic Weights and Isotopic Compositions with Relative Atomic Masses*" by the NIST Physical Measurement Laboratory [1].
This information is embedded in the form of `Isotope` and `Element` classes.
The user can also create custom `Elements`s and specify the isotopic composition manually.
`Molecule`s are made from `Element`s and `Mixture`s can be made from `Element`s, `Molecule`s and other `Mixture`s.
A thorough example is given in Section 'Example'.


## Installation

The source code of the most recent development version is hosted on [GitHub](https://github.com/jmercz/isoVec).
The binary installers and the source code of stable releases is available on the project site on the [Python Package Index (PyPI)](https://pypi.org/project/isovec) and can be simply installed from the repository via

```sh
pip install isovec
```

There are no additional dependencies other than core Python packages of Python 3.10.


## Example

The composition of the atmosphere of the Earth may serve as an example how to use isoVec.
The constituents and their mole fractions are taken from the respective [Wikipedia article](https://en.wikipedia.org/wiki/Atmosphere_of_Earth#Composition) and are as of April 2022:

| Constituent                     | Mole Fraction          |
| ------------------------------- | ---------------------- |
| Nitrogen (N<sub>2</sub>)        | 78.084 %               |
| Oxygen (O<sub>2</sub>)          | 20.946 %               |
| Argon (Ar)                      | 0.9340 %               |
| Carbon dioxide (CO<sub>2</sub>) | 417 ppm                |
| Neon (Ne)                       | 18.18 ppm              |
| Helium (He)                     | 5.24 ppm               |
| Methane (CH<sub>4</sub>)        | 1.87 ppm               |
| Krypton (Kr)                    | 1.14 ppm               |

It may be noted, that these values are often given as a volume fraction in other literature.
For a mixture of ideal gases (which is the case for air), volume and mole fractions are equal.
Furthermore the amount of substance (mole) and number of particles (e.g. atoms) are proportional in an element, being connected via the Avogadro constant $N_\mathrm{A}$.
Here, the terms mole and atomic fraction can be used interchangeably.
In general, and especially when dealing with molecules containing multiple atoms per particle, this is not strictly correct, but is done regardless in the API for the sake of convenience.
One should be aware of this and always consider the current context.

Let's assume for the following, that isoVec is imported with the following (shorter) alias.
Keep in mind, that all content is packed inside the given namespace.

```python
import isovec as iso
```

### Element

Natural occuring elements are already implemented and can be accessed via its symbol and "_nat".
Nitrogen for example is called with `N_nat`.
For the sake of this tutorial, we will define nitrogen again by ourselfs.
Furthermore, an extensive library of isotopes is implemented. Each `Isotope` is acessed via the elements symbol, followed by an underscore "\_" and its mass number (that is protons + neutrons).
Nitrogen-14 is therefore called via `N_14`.
An `Element` can only be composed of its respective isotopes, so `Isotope` objects with the same atomic number `Z`.

A custom `Element` is created with a unique name and its composition as a dictionary:

```python
nitrogen = iso.Element("nitrogen", {
    iso.N_14: 9.963600E-01,
    iso.N_15: 3.640000E-03
})
```

In composition dictionaries, positive values refer to an atomic fraction (as done before), while negative values refer to a weight fraction.
Whatever information is available or more convenient may be used, but atomic and weight fractions **cannot** be mixed inside one composition dictionary.
If the keyword argument `mode` is supplied with either `"atomic"`, `"weight"` or other short forms thereof, signs are ignored and the values of the dictionary are interpreted accordingly, making the construction call more verbose.
Alternatively, fraction-specific constructor methods can be called on the class, namely `from_atomic()` and `from_weight()`, in which case the keyword `mode` doesn't need to be supplied.
Otherwise, the syntax is identical to the ordinary constructor call.
The given composition doesn't necessarily need to add up to unity.
Instead, each fraction is normalised by the total sum of given fractions.

The molar mass of an element is calculated automatically, but could be overwritten when supplying the keyword argument `M=value` in the constructor.
Optionally, the density of the element can be supplied with the keyword argument `rho=value`, since densities of elements cannot be calculated automatically by the given informations.
The described behaviour of the constructur also applies to the following classes, if not stated otherwise.
Furthermore, instances of `Element`s and the following classes can be (deep)copied by calling the `copy()` method on the instance.
Properties like density `rho`, molar mass `M`, `name` and `symbol` can be altered from the original, by supplying these as keyword arguments.


### Molecule

The composition for a `Molecule` is the number of atoms per `Element`.
Therefore, only the `"atomic"` mode is valid, or in short positive values in the composition dictionary.
However, this class isn't strictly limited to molecules, but can be used for crystalline or amorphous structures with a stochiometric distribution.

Using the implemented natural `Element`s, the `Molecule` for carbon dioxide is given by:

```python
carbon_dioxide = iso.Molecule("carbon dioxide", {
    iso.C_nat: 1,
    iso.O_nat: 2
})
```

Likewise, the other `Molecule`s are given as:

```python
methane = iso.Molecule("methane", {
    iso.C_nat: 1,
    iso.H_nat: 4
})
nitrogen2 = iso.Molecule("molecular nitrogen", { iso.N_nat: 2})
oxygen2 = iso.Molecule("molecular oxygen", { iso.O_nat: 2})
```

Note, that the molar mass of the entire molecule is calculated by default.


### Mixture

`Mixtures` can contain several pure `Element`s, `Molecule`s and also other `Mixture`s.
In contrast to `Element` and `Molecule`, it is also possible to supply volume fractions in the composition, provided that all constituents have a density defined.
This can be invoked either by supplying the keyword argument `mode="volume"` to the constructor, or calling the specific constructor `from_volume()`.

According to the table from the introduction, a `Mixture` for air is defined by the following:

```python
air = iso.Mixture("air", {
    nitrogen2:                  78.084E-02,  # Molecule
    oxygen2:        iso.percent(20.946),     # Molecule
    iso.Ar_nat:     iso.percent( 0.9340),    # Element
    carbon_dioxide:            417.0E-06,    # Molecule
    iso.Ne_nat:        iso.ppm( 18.18),      # Element
    iso.He_nat:        iso.ppm(  5.24),      # Element
    methane:           iso.ppm(  1.87),      # Molecule
    iso.Kr_nat:        iso.ppm(  1.14)       # Element
})
```

Note the use of convencience functions for conversion, an extensive list can be found in the source code file "conversion.py"
When all constituents have their density defined, the density of the `Mixture` can also be calculated automatically.

The entered material hierarchy can be printed in a tree-like structure by invoking `print_tree_input()` on any substance we want to inspect:

```python
air.print_tree_input(weight=True)
```

The percentage refers to the fraction in the parent and will sum up to 100 % on their 'sibling' level.
Setting the keywords `weight` or `volume` to `True` also calculates the weight and volume percent of each node respectively (where applicable), in addition to the atomic percentage that is printed by default (but can be deactivated via the keyword `atomic`).
This yields the following (excerpt of the) output:

```
Mixture "air": 28.9660 g/mol
├── Molecule "molecular nitrogen":  78.0775 at.%  |   75.5097 wt.%  |  28.0134 g/mol
│   └── Element "natural nitrogen": 100.0000 at.%  |  100.0000 wt.%  |  14.0067 g/mol
│       ├── Isotope "N-14":  99.6360 at.%  |   99.6102 wt.%  |  14.0031 g/mol
│       └── Isotope "N-15":   0.3640 at.%  |    0.3898 wt.%  |  15.0001 g/mol
├── Molecule "molecular oxygen":  20.9443 at.%  |   23.1372 wt.%  |  31.9988 g/mol
│   └── Element "natural oxygen": 100.0000 at.%  |  100.0000 wt.%  |  15.9994 g/mol
│       ├── Isotope "O-16":  99.7570 at.%  |   99.7290 wt.%  |  15.9949 g/mol
│       ├── Isotope "O-17":   0.0380 at.%  |    0.0404 wt.%  |  16.9991 g/mol
│       └── Isotope "O-18":   0.2050 at.%  |    0.2306 wt.%  |  17.9992 g/mol
├── Element "natural argon":   0.9339 at.%  |    1.2880 wt.%  |  39.9478 g/mol
│   ├── Isotope "Ar-36":   0.3336 at.%  |    0.3004 wt.%  |  35.9675 g/mol
│   ├── Isotope "Ar-38":   0.0629 at.%  |    0.0598 wt.%  |  37.9627 g/mol
│   └── Isotope "Ar-40":  99.6035 at.%  |   99.6399 wt.%  |  39.9624 g/mol
   [...]
```

The format of the fraction values and the physical properties can be given by supplying the keywords `frac_fmt` and `prop_fmt` respectively with a format string.


### Elemental & Isotopic Composition

To obtain the elemental composition, the method `get_elements()` can be called upon a substance, yielding an ordered dictionary with elements and their respective summed fractions.
With the `mode` keyword argument, fractions can be fetched in terms of `"atomic"` or `"weight"`.
For the mixture of air, the following code

```python
element_vector = air.get_elements(mode="atomic")
for element, at_frac in element_vector.items():
    print(f"{element.symbol:>2}: {at_frac:.4E}")
```

yields the elemental composition:

```
 H: 3.7565E-06
He: 2.6315E-06
 C: 2.1036E-04
 N: 7.8428E-01
 O: 2.1080E-01
Ne: 9.1301E-06
Ar: 4.6906E-03
Kr: 5.7251E-07
```

Similarly, `get_isotopes()` yields the isotopic composition of a substance.
Additionally, the keyword argument `use_natural` can be used to represent the isotopic composition of a natural `Element` with a surrogate `Isotope` object.
This surrogate object represents an element with natural occuring abundance.
`use_natural=True` applies this behaviour to all natural elements.
Alternatively, a collection (iterable) of natural `Element` objects can be passed as the argument, limiting the behaviour to those specified.
Note that custom made elements are not subject to that behaviour, unless the keyword argument `natural=True` was specified in their constructor.
The (atomic) isotopic composition of air - considering natural oxygen as one - can be fetched with the following code:

```python
isotope_vector = air.get_isotopes(mode="atomic", use_natural=(iso.O_nat,))
for isotope, at_frac in isotope_vector.items():
    isotope_name = isotope.name.replace("-0", "-nat")  # surrogate Isotopes have a mass number of zero
    print(f"{isotope_name:>6}: {at_frac:.4E}")
```

yields the isotope vector (with all oxygen isotopes, that are part of a natural composition `Element`, being condensed into a surrogate `Isotope` object):

```
   H-1: 3.7560E-06
   H-2: 4.3200E-10
  He-3: 3.5263E-12
  He-4: 2.6315E-06
  C-12: 2.0811E-04
  C-13: 2.2508E-06
  N-14: 7.8143E-01
  N-15: 2.8548E-03
 O-nat: 2.1080E-01
   [...]
```

The composition can also be printed in a tree-hierarchy.
Both `atomic` and `weight` fractions can be fetched and are limited to `Element` and `Isotope` objects.
The following code

```python
air.print_tree_composition(weight=True)
```

yields the composition tree:

```
Mixture "air": 28.9660 g/mol
├── Molecule "molecular nitrogen": 28.0134 g/mol
│   └── Element "natural nitrogen":  78.4281 at.%  |   75.5097 wt.%  |  14.0067 g/mol
│       ├── Isotope   "N-14":  78.1426 at.%  |   75.2154 wt.%  |  14.0031 g/mol
│       └── Isotope   "N-15":   0.2855 at.%  |    0.2943 wt.%  |  15.0001 g/mol
├── Molecule "molecular oxygen": 31.9988 g/mol
│   └── Element "natural oxygen":  21.0383 at.%  |   23.1372 wt.%  |  15.9994 g/mol
│       ├── Isotope   "O-16":  20.9872 at.%  |   23.0744 wt.%  |  15.9949 g/mol
│       ├── Isotope   "O-17":   0.0080 at.%  |    0.0093 wt.%  |  16.9991 g/mol
│       └── Isotope   "O-18":   0.0431 at.%  |    0.0534 wt.%  |  17.9992 g/mol
├────── Element "natural argon":   0.4691 at.%  |    1.2880 wt.%  |  39.9478 g/mol
│       ├── Isotope  "Ar-36":   0.0016 at.%  |    0.0039 wt.%  |  35.9675 g/mol
│       ├── Isotope  "Ar-38":   0.0003 at.%  |    0.0008 wt.%  |  37.9627 g/mol
│       └── Isotope  "Ar-40":   0.4672 at.%  |    1.2834 wt.%  |  39.9624 g/mol
├── Molecule "carbon dioxide": 44.0095 g/mol

CHANGELOG.md

# Changelog

## Version 1.2.1 (14/11/2023)

- added `copy()` function to `Substance`
	- deep-copy of the substance
	- keyword arguments to change properties are:
		- `name` for the descriptive name
		- `symbol` for the symbol
		- `rho` for density
		- `M` for molar mass



## Version 1.2.0.1 (03/11/2023)

- fixed typos in "README.md"



## Version 1.2.0 (03/11/2023)

This update features a major overhaul of the routines to fetch isotopic (and elemental) compositions.
Atomic fractions of `Molecule`s in `Mixture`s cannot be passed down directly.
This would underestimate the atomic fraction of the `Element`s of that `Molecule`.
Instead, all `Element`s have to be collected according to their actual atomic abundance and subsequently be normalised.
When the composition input mode was `weight` and the isotopic composition was fetched in `weight` as well, values were **correct** (because of back and forth conversion).
Update 1.1.1 was ment to adress the faulty isotopic composition, when fetched in ´atomic´ mode, but introduced other errors and is thus discouraged to be used.
The respective releases on PyPI therefore have been yanked.
This new approach also required changes to the tree-printing routine.
`print_tree()` has been replaced by `print_tree_input()`, that acts like the old `print_tree(scale=False)`.
To show the actual (scaled) elemental and isotopic composition, `print_tree_composition()` has to be used.

- added `get_elements()` method to `Substance` for elemental composition of that `Substance`
	- keyword `mode` for either "atomic" or "weight" fractions
- added option for `get_isotopes()` to supply explicit collection of (natural) elements to be considered
	- removed the internal method `_append_isotopes()`
	- instead uses `_isotopic_composition()`, `_elemental_composition()` and `_append_elements()` to get correct (atomic) composition
		- `_isotopic_composition()`, `_elemental_composition()` return a dictionary, matching the position in the tree hierarchy to the `Isotope` or `Element` and their fraction respectively
- added `surrogate_isotope()` method to `Element` to get a surrogate `Isotope` object in case of a natural element (used for `use_natural` algorithm)
- removed `print_overview()` from `Mixture`, `Molecule` and `Element`, because it doesn't comply with the new approach and has been superseded by tree-printing
- `Node`s in the tree use their position in the tree-hierarchy as their id
- replaced `print_tree()` with
	- `print_tree_input()`
		- shows input composition in selected fraction modes (`atomic`, `weight` and `volume`)
		- similar to old `print_tree(scale=False)`
	- `print_tree_composition()`
		- shows scaled composition of `Element`s and `Isotopes`s of a `Substance`
		- possible fraction modes: `atomic` and `weight`
		- uses `Node` id to fetch the correct composition value from `_isotopic_composition()` or `_elemental_composition()` dictionary
- added more sophisticated right alignment options for tree-plotting
	- added `_print_depth()` method for `Node` class, to get desired level of printing that `Node` object
	- new prefix generation in `print_node()` of `Node` class
	- right-alignment-preference class variable for `Substance`
	- `Element`s and `Isotope`s are respectively right aligned in `print_tree_composition()`
- added internal conversion consistency case to "validation.py"
- added detailed air case to "validation.py"



## Version 1.1.1.1 (16/10/2023)

- fixed problem with tree plotting of molecules, that lead to wrong atomic isotope vectors
	- changed molar mass of molecule to average molar mass (as other substances)
	- added molar mass of molecule (is calculated as before)
	- added `_get_data_dict()` to `Substance` (and `Molecule`) for more flexibility in tree plotting
- added option for `get_isotopes()` to supply explicit collection of (natural) elements to be considered


## Version 1.1.1 (16/10/2023)

- fixed problem with molar mass of molecules, that lead to wrong atomic isotope vectors
	- changed molar mass of molecule to average molar mass (as other substances)
	- added molar mass of molecule (is calculated as before)
	- added `_get_data_dict()` to `Substance` (and `Molecule`) for more flexibility in tree plotting
- added option for `get_isotopes()` to supply explicit collection of (natural) elements to be considered


## Version 1.1.0 (18/08/2023)

This updates marks a big overhaul of the code structure.
`Element`, `Molecule` and `Mixture` now inherit from the abstract base class `Substance`.
This will help in terms of future maintainability.
Volume fractions are introduced as a possible input for `Mixture`s.
All modules, classes, methods and constants received docstrings for documentation and a more helpful API.
Now, all isotopes from the NIST database are implemented, increasing the available `Isotope` object count from 357 to 3355.
Added `Node` class for easier access to hierarchical tree structure of substances.
Furthermore, this class allows a tree structure print, that should be clearer than previous print methods.
`Isotope`, `Substance` and all subclasses received various additional attributes and methods for enhanced functionality.


### Isotope

- changes to `Isotope`:
	- added attribute relative atomic mass `A_r`
		- molar mass `M` is calculated via `A_r` and molar mass constant `M_u` (according to 2019 redefinitions)
	- added isomeric state `I`
	- added property `ZA` (notation)
	- added property `ZAI` (notation)
	- name is not longer a positional argument in constructor and is generated automatically, but can be overriden with `name` keyword
	- added `element_symbol()` method to get symbol of associated element
- added all isotopes from [NIST database](https://www.nist.gov/pml/atomic-weights-and-isotopic-compositions-relative-atomic-masses)
	- total amount: 3355 isotopes


### Substance and subclasses

- `Substance` base class for maintainability
	- unified attributes and methods where possible

validation.py

import copy

from tabulate import tabulate

from src import isovec as iso


if __name__ == "__main__":

    print_raw_output = True  # print raw output of script calculations

    # tabulate options
    headers = ["", "expected result", "script result", "deviation"]  # header row for table
    def compare(comparison: str, expected: float, script: float) -> list:
        """Make row for table that compares expected and script results."""
        return [comparison, expected, script, script - expected]
    table_kwargs = {
        "tablefmt": "github",
        "stralign": "right"
    }

    ###################
    ### Ni80Cr20
    ###################
    desc = r"""
    test conversion from weight to atom percent
    as in https://www.plasmaterials.com/converting-atomic-percent-to-weight-percent-and-vice-versa/
    Alloy of 80 wt.% Nickel and 20 wt.% Chromium
    """

    print()
    print(r"Ni80Cr20 Test Case")
    print(desc)

    ## hand calculation

    # conversion formula given on webpage
    atPercX = lambda wtPercX, atWtX, wtPercY, atWtY: (wtPercX/atWtX) / ((wtPercX/atWtX) + (wtPercY/atWtY)) * 100


    ## isovec calculation

    # create copies of elements and overwrite atomic weights to match source
    nickel = copy.deepcopy(iso.Ni_nat)
    nickel._atomicWeight = 58.71
    chromium = copy.deepcopy(iso.Cr_nat)
    chromium._atomicWeight = 51.99

    Ni80Cr20 = iso.Mixture("Ni80Cr20", {
        nickel:   -80e-2,
        chromium: -20e-2
    })

    # extract results
    res_tree = Ni80Cr20.make_node("input")
    wtP_Ni = round(res_tree.get_nodes_by_content(nickel)[0].data["x"]*100, 2)
    wtP_Cr = round(res_tree.get_nodes_by_content(chromium)[0].data["x"]*100, 2)

    res_table = [
        compare("Nickel [at.%]",   round(atPercX(80, 58.71, 20, 51.99), 2), wtP_Ni),
        compare("Chromium [at.%]", round(atPercX(20, 51.99, 80, 58.71), 2), wtP_Cr),
    ]
    print(tabulate(res_table, headers=headers, **table_kwargs))
    print()

    if print_raw_output:
        print(30*"#")
        print()
        print("--- Alloy calculated by script: ---")
        Ni80Cr20.print_tree_input(weight=True)



    print()
    print()
    print(80*"X")
    ###################
    ### Fe-7wt%C
    ###################
    desc = r"""
    test conversion from weight to atom percent
    as in https://www.southampton.ac.uk/~pasr1/g7.htm
    Alloy of 93 wt.% Iron and 7 wt.% Carbon
    """

    print()
    print(r"Fe-7wt%C Test Case")
    print(desc)

    Fe7C = iso.Mixture("Fe-7C", {
        iso.Fe_nat: -93e-2,
        iso.C_nat:   -7e-2
    })

    # extract results
    res_tree = Fe7C.make_node("input")
    wtP_C = round(res_tree.get_nodes_by_content(iso.C_nat)[0].data["x"]*100, 0)

    res_table = [
        compare("Carbon [at.%]", 26, wtP_C),
    ]
    print(tabulate(res_table, headers=headers, **table_kwargs))
    print()

    if print_raw_output:
        print(30*"#")
        print()
        print("--- Alloy calculated by script: ---")
        Fe7C.print_tree_input(weight=True)



    print()
    print()
    print(80*"X")
    ###################
    ### Fe-25at%C
    ###################
    desc = r"""
    test conversion from atom to weight percent
    as in https://www.southampton.ac.uk/~pasr1/g7.htm
    Alloy of 75 at.% Iron and 25 at.% Carbon
    """

    print()
    print(r"Fe-25at%C Test Case")
    print(desc)

    Fe25C = iso.Mixture("Fe-25at.%C", {
        iso.Fe_nat: 75e-2,
        iso.C_nat:  25e-2
    })

    # extract results
    res_tree = Fe25C.make_node("input", weight=True)
    wtP_C = round(res_tree.get_nodes_by_content(iso.C_nat)[0].data["w"]*100, 2)

    res_table = [
        compare("Carbon [wt.%]", 6.66, wtP_C),
    ]
    print(tabulate(res_table, headers=headers, **table_kwargs))
    print()

    if print_raw_output:
        print(30*"#")
        print()
        print("--- Alloy calculated by script: ---")
        Fe25C.print_tree_input(weight=True)



    print()
    print()
    print(80*"X")
    ###################
    ### 98 g Sn - 65 g Pb
    ###################
    desc = r"""
    test conversion from weight to atom percent
    as in https://www.youtube.com/watch?v=0otCqXjZkOw
    Alloy of 98 g tin and 65 g lead
    """

    print()
    print(r"98 g Sn - 65 g Pb Test Case")
    print(desc)

    m_tin = 98  # [g]
    m_lead = 65  # [g]
    m_tot = m_tin + m_lead
    SnPb = iso.Mixture("Sn(98)Pb(65)", {
        iso.Sn_nat: -m_tin/m_tot,
        iso.Pb_nat: -m_lead/m_tot
    })
    SnPb_dir = iso.Mixture("Sn(98)Pb(65)_direct", {
        iso.Sn_nat: -98,
        iso.Pb_nat: -65
    })

    # extract results
    res_tree = SnPb.make_node("input")
    atP_Sn = round(res_tree.get_nodes_by_content(iso.Sn_nat)[0].data["x"]*100, 2)
    atP_Pb = round(res_tree.get_nodes_by_content(iso.Pb_nat)[0].data["x"]*100, 2)

    res_table = [
        compare("Tin [at.%]",  72.47, atP_Sn),
        compare("Lead [at.%]", 27.53, atP_Pb)
    ]
    print(tabulate(res_table, headers=headers, **table_kwargs))
    print()

    if print_raw_output:
        print(30*"#")
        print()
        print("--- Alloy calculated by script: ---")
        SnPb.print_tree_input(weight=True)

        print()
        print("--- Alloy calculated by script with masses entered directly (shows normalisation): ---")
        SnPb_dir.print_tree_input(weight=True)



    print()
    print()
    print(80*"X")
    ###################
    ### 20 vol.% alcohol in water
    ###################
    desc = r"""
    test conversion from volume to weight percent
    as in https://rechneronline.de/volume-mass-percent/
    Mixture of 20 vol.% alcohol in water
    """

    print()
    print(r"20 vol.% alcohol in water Test Case")
    print(desc)

    alcohol = iso.Molecule(
        "alcohol", {},
        M=46.08,
        rho=0.789  # from website
    )
    water = iso.Molecule(
        "water", {},
        M=18.01528,
        rho=1.  # from website
    )
    mix = iso.Mixture("mix", {
        alcohol: 20,
        water: 100-20
    }, mode="volume")

    # extract results
    res_tree = mix.make_node("input", weight=True)
    wtP_alcohol = round(res_tree.get_nodes_by_content(alcohol)[0].data["w"]*100, 1)

    res_table = [
        compare("Alcohol [vol.%]", 16.5, wtP_alcohol)
    ]
    print(tabulate(res_table, headers=headers, **table_kwargs))
    print()

    if print_raw_output:
        print(30*"#")
        print()
        print("--- Mixture calculated by script: ---")
        mix.print_tree_input(weight=True, volume=True)



    print()
    print()
    print(80*"X")
    ###################
    ### Internal composition fetching and conversion
    ###################
    desc = r"""
    test consistency of internal composition fetching and conversion
    """

    print()
    print(r"Internal composition fetching and conversion")
    print(desc)

    carbon_dioxide = iso.Molecule("carbon dioxide", {
        iso.C_nat: 1,
        iso.O_nat: 2
    })
    methane = iso.Molecule("methane", {
        iso.C_nat: 1,
        iso.H_nat: 4
    })
    nitrogen2 = iso.Molecule("molecular nitrogen", { iso.N_nat: 2})
    oxygen2 = iso.Molecule("molecular oxygen", { iso.O_nat: 2})
    air = iso.Mixture("air", {
        nitrogen2:                  78.084E-02,  # Molecule
        oxygen2:        iso.percent(20.946),     # Molecule
        iso.Ar_nat:     iso.percent( 0.9340),    # Element
        carbon_dioxide:            417.0E-06,    # Molecule
        iso.Ne_nat:        iso.ppm( 18.18),      # Element
        iso.He_nat:        iso.ppm(  5.24),      # Element
        methane:           iso.ppm(  1.87),      # Molecule
        iso.Kr_nat:        iso.ppm(  1.14)       # Element
    })
    differences = air._compare_converted_isotopes()

    print()
    dif_at = differences["at"]
    i_max_dif_at = dif_at.index(max(dif_at, key=abs))
    print(f"Maximum atomic difference: {dif_at[i_max_dif_at]} (" + str(differences["rel_at"][i_max_dif_at]) + ")")
    dif_wt = differences["wt"]
    i_max_dif_wt = dif_wt.index(max(dif_wt, key=abs))
    print(f"Maximum weight difference: {dif_wt[i_max_dif_wt]} (" + str(differences["rel_wt"][i_max_dif_wt]) + ")")



    print()
    print()
    print(80*"X")
    ###################
    ### Detailed air composition
    ###################
    desc = r"""
    test conversion from atomic to weight percent for air
    as in https://de.wikipedia.org/wiki/Luft#Zusammensetzung
    """

    print()
    print(r"Detailed air composition")
    print(desc)

    N2 = iso.Molecule("molecular nitrogen", {iso.N_nat: 2})
    O2 = iso.Molecule("molecular oxygen",   {iso.O_nat: 2})
    H2 = iso.Molecule("molecular hydrogen", {iso.H_nat: 2})
    CO2 = iso.Molecule("carbon dioxide", {
        iso.C_nat:  1,
        iso.O_nat:  2
    }, symbol="CO2")
    CH4 = iso.Molecule("methane", {
        iso.C_nat:  1,
        iso.H_nat:  4
    }, symbol="CH4")
    N2O = iso.Molecule("dinitrogen oxide", {
        iso.N_nat:  2,
        iso.O_nat:  1
    }, symbol="N2O")
    CO = iso.Molecule("carbon monoxide", {
        iso.C_nat:  1,
        iso.O_nat:  1
    }, symbol="CO")
    CCl2F2 = iso.Molecule("dichlorodifluoromethane", {
        iso.C_nat:  1,
        iso.Cl_nat: 2,
        iso.F_nat:  2
    }, symbol="CCl2F2")
    CCl3F = iso.Molecule("trichlorofluoromethane", {
        iso.C_nat:  1,
        iso.Cl_nat: 3,
        iso.F_nat:  1
    }, symbol="CCl3F")
    CHClF2 = iso.Molecule("chlorodifluoromethane", {
        iso.C_nat:  1,
        iso.H_nat:  1,
        iso.Cl_nat: 1,
        iso.F_nat:  2
    }, symbol="CHClF2")
    CCl4 = iso.Molecule("tetrachloromethane", {
        iso.C_nat:  1,
        iso.Cl_nat: 4,
    }, symbol="CCl4")
    CClF2CCl2F = iso.Molecule("1,1,2-trichloro-1,2,2-trifluoroethane", {
        iso.C_nat:  2,
        iso.Cl_nat: 3,
        iso.F_nat:  3,
    }, symbol="CClF2CCl2F")
    C2H3Cl2F = iso.Molecule("1,1-dichloro-1-fluoroethane", {
        iso.C_nat:  2,
        iso.H_nat:  3,
        iso.Cl_nat: 2,
        iso.F_nat:  1,
    }, symbol="C2H3Cl2F")
    C2H3ClF2 = iso.Molecule("1-chloro-1,1-difluoroethane", {
        iso.C_nat:  2,
        iso.H_nat:  3,
        iso.Cl_nat: 1,
        iso.F_nat:  2,
    }, symbol="C2H3ClF2")
    SF6 = iso.Molecule("sulphur hexafluoride", {
        iso.S_nat:  1,
        iso.F_nat:  6,
    }, symbol="SF6")
    CBrClF2 = iso.Molecule("bromochlorodifluoromethane", {
        iso.C_nat:  1,
        iso.Br_nat: 1,
        iso.Cl_nat: 1,
        iso.F_nat:  2,
    }, symbol="CBrClF2")
    CBrF3 = iso.Molecule("bromotrifluoromethane", {
        iso.C_nat:  1,
        iso.Br_nat: 1,
        iso.F_nat:  3,
    }, symbol="CBrF3")

    air = iso.Mixture("dry air", {
        # main components
        N2:         iso.percent(78.084),
        O2:         iso.percent(20.942),
        iso.Ar_nat: iso.percent( 0.934),
        # trace gases
        CO2:        iso.ppm(400.0),
        iso.Ne_nat: iso.ppm( 18.18),
        iso.He_nat: iso.ppm(  5.24),
        CH4:        iso.ppm(  1.85),
        iso.Kr_nat: iso.ppm(  1.14),

        H2:         iso.ppb(500.0),
        N2O:        iso.ppb(328.0),
        CO:         iso.ppb(175.0),
        iso.Xe_nat: iso.ppb( 87.0),

        CCl2F2:     iso.ppt(520.0),
        CCl3F:      iso.ppt(234.0),
        CHClF2:     iso.ppt(253.0),
        CCl4:       iso.ppt( 81.0),
        CClF2CCl2F: iso.ppt( 71.0),
        C2H3Cl2F:   iso.ppt( 26.0),
        C2H3ClF2:   iso.ppt( 23.0),
        SF6:        iso.ppt(  8.0),
        CBrClF2:    iso.ppt(  4.0),
        CBrF3:      iso.ppt(  3.4),
    }, mode="atomic")

    # extract results
    res_tree = air.make_node("input", weight=True)
    res_table = [
        compare("N2 [wt.%]", 75.518, round(res_tree.get_nodes_by_content(N2)[0].data["w"]*1e2, 3)),
        compare("O2 [wt.%]", 23.135, round(res_tree.get_nodes_by_content(O2)[0].data["w"]*1e2, 3)),
        compare("Ar [wt.%]",  1.288, round(res_tree.get_nodes_by_content(iso.Ar_nat)[0].data["w"]*1e2, 3)),
        [],
        compare("CO2 [wt.ppm]",  590, round(res_tree.get_nodes_by_content(CO2)[0].data["w"]*1e6, 0)),
        compare("Ne [wt.ppm]",    12.67, round(res_tree.get_nodes_by_content(iso.Ne_nat)[0].data["w"]*1e6, 2)),
        compare("He [wt.ppm]",     0.72, round(res_tree.get_nodes_by_content(iso.He_nat)[0].data["w"]*1e6, 2)),
        compare("CH4 [wt.ppm]",    0.97, round(res_tree.get_nodes_by_content(CH4)[0].data["w"]*1e6, 2)),
        compare("Kr [wt.ppm]",     3.30, round(res_tree.get_nodes_by_content(iso.Kr_nat)[0].data["w"]*1e6, 2)),
        [],
        compare("H2 [wt.ppb]",    36, round(res_tree.get_nodes_by_content(H2)[0].data["w"]*1e9, 0)),
        compare("N2O [wt.ppb]",  480, round(res_tree.get_nodes_by_content(N2O)[0].data["w"]*1e9, 0)),
        compare("CO [wt.ppb]",   175, round(res_tree.get_nodes_by_content(CO)[0].data["w"]*1e9, 0)),
        compare("Xe [wt.ppb]",   400, round(res_tree.get_nodes_by_content(iso.Xe_nat)[0].data["w"]*1e9, 0)),
        [],
        compare("CCl2F2 [wt.ppt]",    2200, round(res_tree.get_nodes_by_content(CCl2F2)[0].data["w"]*1e12, 0)),
        compare("CCl3F [wt.ppt]",     1100, round(res_tree.get_nodes_by_content(CCl3F)[0].data["w"]*1e12, 0)),
        compare("CHClF2 [wt.ppt]",     480, round(res_tree.get_nodes_by_content(CHClF2)[0].data["w"]*1e12, 0)),
        compare("CCl4 [wt.ppt]",       510, round(res_tree.get_nodes_by_content(CCl4)[0].data["w"]*1e12, 0)),
        compare("CClF2CCl2F [wt.ppt]", 520, round(res_tree.get_nodes_by_content(CClF2CCl2F)[0].data["w"]*1e12, 0)),
        compare("C2H3Cl2F [wt.ppt]",    70, round(res_tree.get_nodes_by_content(C2H3Cl2F)[0].data["w"]*1e12, 0)),
        compare("C2H3ClF2 [wt.ppt]",    50, round(res_tree.get_nodes_by_content(C2H3ClF2)[0].data["w"]*1e12, 0)),
        compare("SF6 [wt.ppt]",         25, round(res_tree.get_nodes_by_content(SF6)[0].data["w"]*1e12, 0)),
        compare("CBrClF2 [wt.ppt]",     25, round(res_tree.get_nodes_by_content(CBrClF2)[0].data["w"]*1e12, 0)),
        compare("CBrF3 [wt.ppt]",       13, round(res_tree.get_nodes_by_content(CBrF3)[0].data["w"]*1e12, 0)),
    ]
    print(tabulate(res_table, headers=headers, **table_kwargs))
    print()

    if print_raw_output:
        print(30*"#")
        print()
        print("--- Mixture calculated by script: ---")
        air.print_tree_input(weight=True)
